### ISSUE TITLE
Improve reporting of the tactics applied by auto, Ltac, etc.

### ISSUE THREAD

**Issue by jfehrle** [coq/coq#10313]
There are currently a couple ways for a user to get a list of simpler tactics equivalent to what a complex tactic such as `auto` or an Ltac tactic does.  This is valuable for debugging and to users for their understanding, including beginners.  However, the list of tactics in the explanations is not always correct.  This seems like a basic capability that users want in a proof assistant.

Here's a summary of the discussion we had at the Coq Users and Developers Workshop discussed what it would take to improve the current feature.  Plus some further thoughts. 
 Corrections/amplifications appreciated.

Getting the raw data
- `auto` uses a different version of apply than the regular `apply` that may not behave the same when run separately (the list of tactics from `info_auto` says `simple apply`)   (Right??)
- backtracking in `auto` or Ltac doesn't undo all side effects (e.g. resolving an evar), so running the reported list of tactics may leave that unresolved.
- `auto` and Ltac use different versions of 2nd order unification than the regular commands, which may give different answers or one may fail when the other succeeds.
- @MSoegtropIMC uses `Info` to debug his Ltac scripts.  Sometimes the reported list of tactics is incorrect (probably the apply differences) and it takes a long time to figure out a list of tactics that works
- the implementation of `auto` is a bit of a mess and would be difficult to significantly enhance.  IIRC the recommendation is that users should use `typeclasses eauto` going forward--though it would not be 100% compatible for users.
- currently it's not possible to convert every proof term into an equivalent list of tactics.  We could in principle add additional tactics/syntax to allow this.  However, this would tend to generate simple tactics: an `omega` invoked by Ltac would only appear as a series of simpler tactics.

User interface
- the reported list of tactics isn't formatted exactly as you'd pass to Coq.  Ex "simple apply or_intror (in core)", also `split` sometimes comes out with extra characters around it.
- there are multiple ways to get reports back.  These should be made consistent.  Ex: `info_auto` vs `Info` vs `Set Info Level`.  Not sure if `info_trivial` is similar.
- we should consider whether to rewrite the raw list of tactics to make it more similar to what a user would type.  For example, should "intro; intro; intro; intro; intro;" appear as "intros"?  Should we add bullets when there are multiple subgoals?  Perhaps it would be useful to have finer-grained controls for Ltac, e.g. only trace particular tactics.
- there should be a section in the documentation describing these reporting features and how to use them.  `info_auto` is not well documented.  It's not listed in the tactics index.
- the feature doesn't have to be perfect to be useful.  But the limitations should be mentioned in the doc so users know what to expect.

For working on this feature, it would be helpful to have instrumentation that could run through a large corpus of scripts and identify cases where the reported list of simpler tactics isn't equivalent to `auto` or Ltac.

We should have a medium-term goal to have these complex tactics always report an equivalent list of tactics.  Some of the user interface improvements could be done immediately.

Participants in the discussion: @herbelin @ppedrot @Zimmi48 @MSoegtropIMC @jfehrle.

**Comment by ejgallego** [comment]
I think there may be some related issues, would be great to triage them an unify.

**Comment by jfehrle** [comment]
Can you give some specifics?  Otherwise I have no idea what you have in mind.

**Comment by ejgallego** [comment]
> Can you give some specifics? Otherwise I have no idea what you have in mind.

I mean there already exists some reported issues about problem tactic / trace debug, I believe.

**Comment by herbelin** [comment]
Hi @jfehrle, I would say that nothing is impossible to achieve per se among what you are listing, but, as we discussed live with @ppedrot @Zimmi48 and @MSoegtropIMC, there are intertwined issues and that would be a pity to work on some plan which for some reasons gets obsolete after some while.

What is the future of `Info`? IIRC, @ppedrot was not so enthousiastic about it but I did not see what alternatives he proposed. Even if `Info` is not 100% correct, wouldn't it be worth? I'm ready to spend a bit on time on improving `Info` but that would be a pity to do so if it is to break it eventually.

In particular, if `Info` is where we put our energy, and `typeclasses eauto` becomes the main `auto` engine, there is probably no need to port `info_auto` to `typeclasses auto`. Otherwise, if we cannot rely on `Info` working enough in a short term, I guess that the best way shall be to adapt  `info_auto` to `typeclasses auto` (knowing that there are questions about depth vs levels, etc. but I don't think that these are unsurmountable once it is clear that we really want to address them; at worst with compatibility code).

So, using a familiar terminology, I'm _summoning_ @ppedrot to answer about `Info` (can we rely on it and decide that it is worth and maintain it?), and @mattam82 about the old proof engine (do you have an estimation of what to do to eventually get rid of `logic.ml`? How do you think we can coordinate towards this direction? How can each of us help?).

Once we have a basic infrastructure, we can start working at fixing the printing bugs, documenting the mechanism, applying optimizations (such as contracting `intro; intro` into `do 2 intro`, etc.).

> `auto` uses a different version of apply than the regular `apply` that may not behave the same when run separately (the list of tactics from `info_auto` says `simple apply`)   (Right??)

This is right: `simple apply` is the user-level variant of `apply` which is closest to the `apply` of `auto` but it should be checked though whether this is exactly it - for `auto` - or not exactly it. In the latter case, it would however not be difficult to provide a user-access to this variant, say `apply_for_auto foo` or `apply #[auto] foo`.

> backtracking in auto or Ltac doesn't undo all side effects (e.g. resolving an evar), so running the reported list of tactics may leave that evar unresolved.

I could not understand why it would be so when @ppedrot explained. @ppedrot, can you explain or point to the part of the code which does not backtrack evars?

**Comment by jfehrle** [comment]
Yeah, we should come up with a plan.  Let's talk today.  There are other tactics that should report the list of tactics that were applied, for example `inversion`.  Indeed, why shouldn't almost every tactic do this? 
 We should figure how hard this should be and what could be shared between them.  But I think we'd focus on `auto` (or its replacement) and Ltac first.

For the backtracking bug (if it is indeed a bug) I'd look at fixing the bug while providing a backward compatibility flag to turn on the old behavior for a release or two.  It's very bad to have behavior that is almost impossible to describe or reproduce.  Eh @ppedrot?

(Alternative backtracking idea: perhaps we could detect and preserve the problematic existential variable  OR recover them from the proof term?  But could be a fair bit of work/extra complexity to create a workaround.  Generally better to fix the root cause.)
